Core/include/Acts/EventData/MultiTrajectory.hpp

```diff
@@ -51,136 +53,83 @@ namespace detail_lt {
-      GrowableColumns<Eigen::Array<Scalar, Size, Eigen::Dynamic, Flags>,
-                      SizeIncrement>;
-  using StorageCovariance =
-      GrowableColumns<Eigen::Array<Scalar, Size * Size, Eigen::Dynamic, Flags>,
-                      SizeIncrement>;
 };
+}  // namespace detail_lt
 
-struct IndexData {
-  using IndexType = uint16_t;
-
-  static constexpr IndexType kInvalid = UINT16_MAX;
-
-  IndexType irefsurface = kInvalid;
-  IndexType iprevious = kInvalid;
-  IndexType ipredicted = kInvalid;
-  IndexType ifiltered = kInvalid;
-  IndexType ismoothed = kInvalid;
-  IndexType ijacobian = kInvalid;
-  IndexType iprojector = kInvalid;
-
-  double chi2 = 0;
-  double pathLength;
-  TrackStateType typeFlags;
+// This is public
+template <size_t M, bool ReadOnly = true>
+struct TrackStateTraits {
+  using Parameters =
+      typename detail_lt::Types<eBoundSize, ReadOnly>::CoefficientsMap;
+  using Covariance =
+      typename detail_lt::Types<eBoundSize, ReadOnly>::CovarianceMap;
+  using Measurement = typename detail_lt::Types<M, ReadOnly>::CoefficientsMap;
+  using MeasurementCovariance =
+      typename detail_lt::Types<M, ReadOnly>::CovarianceMap;
+
+  using IndexType = std::uint16_t;
+  static constexpr IndexType kInvalid = std::numeric_limits<IndexType>::max();
+  static constexpr IndexType kNoPrevious = kInvalid - 1;
 
-  IndexType iuncalibrated = kInvalid;
-  IndexType icalibrated = kInvalid;
-  IndexType icalibratedsourcelink = kInvalid;
-  IndexType measdim = 0;
+  constexpr static auto ProjectorFlags = Eigen::RowMajor | Eigen::AutoAlign;
+  using Projector =
+      Eigen::Matrix<typename Covariance::Scalar, M, eBoundSize, ProjectorFlags>;
 };
 
+namespace detail_lt {
+
 /// Proxy object to access a single point on the trajectory.
 ///
 /// @tparam SourceLink Type to link back to an original measurement
 /// @tparam M         Maximum number of measurement dimensions
 /// @tparam ReadOnly  true for read-only access to underlying storage
-template <size_t M, bool ReadOnly = true>
+template <typename trajectory_t, size_t M, bool ReadOnly = true>
 class TrackStateProxy {
  public:
-  using Parameters = typename Types<eBoundSize, ReadOnly>::CoefficientsMap;
-  using Covariance = typename Types<eBoundSize, ReadOnly>::CovarianceMap;
-  using Measurement = typename Types<M, ReadOnly>::CoefficientsMap;
-  using MeasurementCovariance = typename Types<M, ReadOnly>::CovarianceMap;
+  using Parameters = typename TrackStateTraits<M, ReadOnly>::Parameters;
+  using Covariance = typename TrackStateTraits<M, ReadOnly>::Covariance;
+  using Measurement = typename TrackStateTraits<M, ReadOnly>::Parameters;
+  using MeasurementCovariance =
+      typename TrackStateTraits<M, ReadOnly>::Covariance;
+
+  using IndexType = typename TrackStateTraits<M, ReadOnly>::IndexType;
+  static constexpr IndexType kInvalid = TrackStateTraits<M, ReadOnly>::kInvalid;
+  static constexpr IndexType kNoPrevious =
+      TrackStateTraits<M, ReadOnly>::kInvalid;
```

Comment:
That was a tricky bug to find. the kNoPrevious here is defined differently than kNoPreovius in TrackStateTraits. 
In any case, kInvalid seems sufficient.